lib/util/result.ts

```diff
@@ -277,6 +273,10 @@ export class Result<T, E = Error> {
     try {
       const res = fn(this.res.val);
 
+      if (res instanceof Result) {
+        return res;
+      }
+
       if (res instanceof Promise) {
```

Comment:
what if it's an AsyncResult? shouldn't we return it directly?

Reply:
That was my original idea for this PR, but I relalized that since `AsyncResult` is the subclass of `Promise`, this case handles both cases without additional code

Reply:
but we wrap the async results twice. can't this cause unwanted side effects?

Reply:
There is a little bit of overhead during re-wrapping, but I think it's okay